### ISSUE TITLE
Azure not supported for WebUI PKCE Auth Code Flow

### ISSUE THREAD

**Issue by Kerwood** [argoproj/argo-cd#16649]
## Describe the bug

In the upcoming `v2.10.0-rc1` release, @Marvin9 added support for using authorization code flow with PKCE in the webUI, which is lovely. :heart: 
 - https://github.com/argoproj/argo-cd/pull/15889

Unfortunately Azure is non-compliant with RFC8414 standard because the `/.well-known/openid-configuration` endpoint is missing the `code_challenge_methods_supported` property. 
The ArgoCD login flow is stopped checking for that specific property here.
https://github.com/argoproj/argo-cd/blob/3224102664b747e88e7330e59b3b44bc9c28cd66/ui/src/app/login/components/utils.ts#L76-L80

Other projects seem to have the same issue, like kubelogin.
 - https://github.com/int128/kubelogin/issues/858

This was brought to the attention of Microsoft in January 2021 and I do not think that they intend to do anything about it.
 - https://learn.microsoft.com/en-us/answers/questions/218113/openid-connect-authorization-code-flow-with-proof

## Proposal
 Maybe have a setting something like, `forcePKCECodeChallengeMethod: S256` that would skip the `code_challenge_methods_supported` property check ?

**Comment by Kerwood** [comment]
@michael-basil

**Comment by michael-basil** [comment]
I have verified the same issue and agree with the suggestion of a flag to mindfully force the flow.

🙏

**Comment by Marvin9** [comment]
What if we know in the frontend itself if IdP is Azure then we by default skip the error assuming that PKCE still works even if `/.well-known/openid-configuration` omit `code_challenge_methods_supported`. I don't see benefit of additional flag for exception like this. Any thoughts on it?

**Comment by michael-basil** [comment]
It would seem this was debated a fair amount with the kubelogin project (dropping the link again):

* https://github.com/int128/kubelogin/issues/858

That project chose to go with: "--oidc-use-pkce"

To follow this convention that others in the industry may be conditioned for the following flag (or similar) would seem to follow suit:

* skipPKCECodeChallengeMethodVerification

**Comment by Kerwood** [comment]
> What if we know in the frontend itself if IdP is Azure then we by default skip the error assuming that PKCE still works even if `/.well-known/openid-configuration` omit `code_challenge_methods_supported`. I don't see benefit of additional flag for exception like this. Any thoughts on it?

Personally I think that creating a snowflake solution for a specific provider is a bad idea. This could end up with all sorts of snowflake code for different IDP's. I think it would be better to create a generic setting.

@michael-basil are you proposing that the setting shoud be `skipPKCECodeChallengeMethodVerification` ?

**Comment by michael-basil** [comment]
I am suggesting the same and with the name to skip verification or would be simpler in configuration management over time.

For example, we test before we introduce new Argo versions, but the typical engineer wouldn't know the details of PKCE.

They would know that it doesn't work though and then we really are in the weeds at that point anyway.

So I am partial to the basic approach of kubelogin though not fixed on the name, just what came naturally.

**Comment by michael-basil** [comment]
Implementation of both flags as optional.

One flag to skip verification and one flag to force a method ...

This is the most robust and durable after thinking it through further.

**Comment by Kerwood** [comment]
"Skip code challenge verification" sounds good to me.

**Comment by michael-basil** [comment]
Wouldn't the proof of key code exchange be the challenge?

This is skipping the method verification.

Perhaps I am understanding this differently?

**Comment by Kerwood** [comment]
tl;dr
I think either
 - there should not be a frontend check for `code_challenge_methods_supported` property
 - or there should be a setting to skip the check
 - or there should be a setting to set the challenge method manually and therefore the check is skipped.

**Comment by michael-basil** [comment]
In a local build of ArgoCD I have been attempting to see if the PKCE flow will actually work when this method verification check is skipped.

Essentially I have commented out the following in utils.ts

``` ts
    // TODO - https://github.com/argoproj/argo-cd/issues/16649 - this check fails for Azure AD
    //if (!authorizationServer?.code_challenge_methods_supported?.includes('S256')) {
    //    throw new PKCELoginError('Authorization Server does not support S256 code challenge method');
    //}
```

I have configured argocd-cm as:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  labels:
    app.kubernetes.io/name: argocd-cm
    app.kubernetes.io/part-of: argocd
  name: argocd-cm
  namespace: argocd
data:
  admin.enabled: "false"
  url: http://localhost:4000
  oidc.config: |
    name: Azure Identity
    enablePKCEAuthentication: true
    issuer: https://login.microsoftonline.com/<tenant-id>/v2.0
    clientID: <client-id>
    requestedIDTokenClaims:
      groups:
        essential: true
    requestedScopes:
      - openid
      - email
```

And argocd-rbac-cm as:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  labels:
    app.kubernetes.io/name: argocd-rbac-cm
    app.kubernetes.io/part-of: argocd
  name: argocd-rbac-cm
  namespace: argocd
data:
  policy.csv: |
    g, group-of-admins, role:admin
    g, user@company.com, role:admin
  policy.default: ""
  scopes: '[roles, email]'
```

The PKCE flow seems to work fine and a JWT token is on the client called `argocd.token` as I would expect.

However, the _User Info_ screen shows that the user is not logged in and the _Log In_ button keeps re-appearing.

So, it seems that the RBAC configuration is not applied correctly, but I don't see how to debug it (yet).

I have tested the RBAC configuration prior with a non-PKCE OIDC flow without issues, so I am a bit puzzled.

Curious if others have gotten further or would see anything obviously wrong from their Azure configurations.

**Comment by Marvin9** [comment]
@michael-basil can you enable auth in api server and then run? You can either use `ARGOCD_E2E_DISABLE_AUTH` environment variable or use `--disable-auth` flag.

**Comment by michael-basil** [comment]
The flag is set to false in the Makefile:

`ARGOCD_E2E_DISABLE_AUTH=false \`

Are you suggesting to set the value to true.  Seems counterintuitive on first glance.

I figure when I can see the debug messages I will be able to trace through the code to see where it's going haywire.

**Comment by Marvin9** [comment]
What I meant is set DISABLE_AUTH to false. 

I usually don't run that way so I might be wrong. But what I see in Makefile is `ARGOCD_E2E_DISABLE_AUTH=false \` in `start-e2e-local`. Since you run `make start`, you would need to either set false in [global](https://github.com/argoproj/argo-cd/blob/master/Makefile#L50) variable used in [`run-in-test-server`](https://github.com/argoproj/argo-cd/blob/master/Makefile#L89).

**Comment by michael-basil** [comment]
Incremental positive progress has been made ... thanks for the pointer(s) @Marvin9!

**The Good:**

After what seems to be re-enabling client authentication for argo-cd api server, the login, user-info and logout flows through the Web UI work properly.

**The Curious:**

The following configurations in Azure appear necessary for this flow to work, assuming that would be expected:

Web Redirect Uri:

* http://localhost:4000/auth/callback

Single-page application Redirect URIs:

* http://localhost:4000/pkce/verify

**Blockages:**

1. It would be highly beneficial to see the api-server omit DEBUG messages, so a pointer on how to get this to work would be appreciated.  I have tried setting `server.log.level: "debug"` within the config-map `argocd-cmd-params-cm` to no avail.

2. For some reason I cannot create a fresh application for ArgoCD (attempting to create an instance of guestbook via `https://github.com/argoproj/argocd-example-apps.git`), receiving the following error (though it seems likely to be unrelated to this issue and more likely related to an awkward local setup I am still smoothing out)

* Unable to create application: application spec for Test is invalid: InvalidSpecError: Unable to generate manifests in guestbook: rpc error: code = Unknown desc = failed to initialize repository resources: rpc error: code = Internal desc = Failed to initialize git repo: stat <path to cached source> permission denied

Looking promising overall.   Once unblocked above I would be inclined to take a deeper look to see about wiring up example configuration option(s) via code (draft pull request perhaps) per the trending conversation here.  I would be a first time contributor so I make this statement with some optimism and hesitation for commitment!  (Once this would settle I would have to work through the contributor guidelines in finer detail of course for such a PR)

Also, if anyone comes by this and happens to be further along I would be happy to meet on a Slack call to trade notes!

Happy Holidays!

**Comment by michael-basil** [comment]
Building a fresh image using `make image` and deploying to a remote Kubernetes cluster verifies that there is no issue creating applications.

Looking promising overall!

**Comment by Kerwood** [comment]
@michael-basil
I don't mean to be rude but you are getting a bit off topic here. You are posting a lot of information that doesn't really have anything to do with the main issue. Please keep your posts relevant.

The main issue/discussion here is: PKCE auth flow in not working with Azure because the `/.well-known/openid-configuration` endpoint does not support the `code_challenge_methods_supported` property. Which setting should we implement to skip the `.includes('S256')` check. 

Here's a few proposals.
- there should not be a frontend check for `code_challenge_methods_supported` property
- or there should be a setting to skip the check
- or there should be a setting to set the challenge method manually and therefore the check is skipped.

**Comment by michael-basil** [comment]
@Kerwood in response to your suggestion I deleted a couple of my posts (as GitHub comments can be refactored):

**TL/DR:**

* I support the proposals you have outlined
* I have verified the viability of the first proposal described as _there should not be a frontend check_ with an Azure IDP:
  * Running argocd and Kubernetes locally from my machine (via `make start`)
  * Running argocd as an image deployed to a remote Kubernetes cluster aka (via `make image`)

In alignment with the code contribution sentiment of [sometimes it's helpful to have some PoC code along with a proposal](https://argo-cd.readthedocs.io/en/stable/developer-guide/code-contributions/) I am working towards a draft pull request to support the described proposal(s) to help bolster the case.

If anyone would like to collaborate to pair on this coding effort please reach out in CNCF Slack by DM or [#argo-contributors](https://cloud-native.slack.com/archives/C020XM04CUW) so that the commentary here is not confusing for others.

**Comment by michael-basil** [comment]
I tested this fix off of the master branch.  It worked well!   Would someone be able to point me to determine when this would go into a release?
